Interpolater.C

```diff
@@ -0,0 +1,181 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description: Implementation of the Interpolater algorithm.
+
+#include "Interpolater.h"
+
+#include <limits.h>
+#include <cmath>
+#include "Matrix.h"
+#include "scalapack_wrapper.h"
+#include "mpi.h"
+
+/* Use C++11 built-in shared pointers if available; else fallback to Boost. */
+#if __cplusplus >= 201103L
+#include <memory>
+#else
+#include <boost/shared_ptr.hpp>
+#endif
+
+using namespace std;
+
+namespace CAROM {
+
+Interpolater::Interpolater(std::vector<Vector*> parameter_points,
+                           std::vector<Matrix*> rotation_matrices,
+                           int ref_point,
+                           double epsilon)
+{
+    CAROM_VERIFY(parameter_points.size() == rotation_matrices.size());
+    CAROM_VERIFY(parameter_points.size() > 0);
+
+    // Get the rank of this process, and the number of processors.
+    int mpi_init;
+    MPI_Initialized(&mpi_init);
+    if (mpi_init == 0) {
+        MPI_Init(nullptr, nullptr);
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &d_rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &d_num_procs);
+
+    d_parameter_points = parameter_points;
+    d_rotation_matrices = rotation_matrices;
+    d_ref_point = ref_point;
+    d_lambda_T = NULL;
+    d_epsilon = epsilon;
+}
+
+std::vector<double> Interpolater::obtainRBF(Vector* point)
+{
+    // Inverse quadratic RBF
+    std::vector<double> inv_q;
+    for (int i = 0; i < d_parameter_points.size(); i++)
+    {
+        Vector diff;
+        point->minus(*d_parameter_points[i], diff);
+        double res = 1.0 / (1.0 + std::pow(d_epsilon*diff.norm(), 2));
+        inv_q.push_back(res);
+    }
+
+    return inv_q;
+}
+
+std::vector<Matrix*> obtainRotationMatrices(std::vector<Vector*> parameter_points,
+        std::vector<Matrix*> bases,
+        int ref_point)
+{
+    // Get the rank of this process, and the number of processors.
+    int mpi_init;
+    int rank;
+    int num_procs;
+    MPI_Initialized(&mpi_init);
+    if (mpi_init == 0) {
+        MPI_Init(nullptr, nullptr);
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
+
+    std::vector<Matrix*> rotation_matrices;
+
+    // Obtain the rotation matrices to rotate the bases into
+    // the same generalized coordinate space.
+    for (int i = 0; i < parameter_points.size(); i++)
+    {
+        CAROM_VERIFY(bases[i]->numRows() == bases[ref_point]->numRows());
+        CAROM_VERIFY(bases[i]->numColumns() == bases[ref_point]->numColumns());
```

Comment:
What about the case that different parameters have different basis dimensions?

Reply:
I'm not sure how that would work since we need to multiply the bases together, and also need to multiply the reduced matrices together, so they would have to be the same dimensions

Reply:
yep. they need to be the same dimension.

Reply:
The dimension determined by an energy fraction usually varies for different parameters. In that case, is it necessary to take the maximum of all dimensions, in order to use this interpolation method?

Reply:
Yep, we have to use the maximum of all dimensions. Instead of energy fraction criteria, there is an option for specifying a specific number for the reduced space dimension, right?

Reply:
Yes, that is better. I'll change to do that. Previously, I was ensuring the basis dimensions had the same reduced dimension beforehand.

Reply:
We need to specify the reduced space dimension and not use energy fraction criteria (which is what I currently instruct the user to do when running the problem). Ignore my previous reply.